smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/directed/DirectedCodegenRunner.java

```diff
@@ -0,0 +1,425 @@
+        LOGGER.fine(() -> "Creating a codegen context for " + directedCodegen.getClass().getName());
+        return directedCodegen.createContext(new CreateContext<>(
+                model, settings, serviceShape, provider, fileManifest));
+    }
+
+    private void registerInterceptors(C context, List<I> integrations) {
+        LOGGER.fine(() -> "Registering CodeInterceptors from integrations of " + integrationClass.getName());
+        List<CodeInterceptor<? extends CodeSection, W>> interceptors = new ArrayList<>();
+        for (I integration : integrations) {
+            interceptors.addAll(integration.interceptors(context));
+        }
+        context.writerDelegator().setInterceptors(interceptors);
+    }
+
+    private void generateShapesInService(C context, ServiceShape serviceShape, Set<Shape> shapes) {
+        LOGGER.fine(() -> "Generating shapes for " + directedCodegen.getClass().getName());
+        generateResourceShapes(context, serviceShape, shapes);
+        generateStructures(context, serviceShape, shapes);
+        generateUnionShapes(context, serviceShape, shapes);
+        generateEnumShapes(context, serviceShape, shapes);
+        LOGGER.finest(() -> "Finished generating shapes for " + directedCodegen.getClass().getName());
+    }
+
+    private void generateResourceShapes(C context, ServiceShape serviceShape, Set<Shape> shapes) {
+        for (ResourceShape shape : model.getResourceShapes()) {
+            if (shapes.contains(shape)) {
+                LOGGER.finest(() -> "Generating resource " + shape.getId());
+                directedCodegen.generateResource(
+                        new GenerateResource<>(context, serviceShape, shape));
+            }
+        }
+    }
+
+    private void generateStructures(C context, ServiceShape serviceShape, Set<Shape> shapes) {
+        for (StructureShape shape : model.getStructureShapes()) {
+            if (shapes.contains(shape)) {
+                if (shape.hasTrait(ErrorTrait.class)) {
+                    LOGGER.finest(() -> "Generating error " + shape.getId());
+                    directedCodegen.generateError(new GenerateError<>(context, serviceShape, shape));
+                } else {
+                    LOGGER.finest(() -> "Generating structure " + shape.getId());
+                    directedCodegen.generateStructure(new GenerateStructure<>(context, serviceShape, shape));
+                }
+            }
+        }
+    }
+
+    private void generateUnionShapes(C context, ServiceShape serviceShape, Set<Shape> shapes) {
+        for (UnionShape shape : model.getUnionShapes()) {
+            if (shapes.contains(shape)) {
+                LOGGER.finest(() -> "Generating union " + shape.getId());
+                directedCodegen.generateUnion(new GenerateUnion<>(context, serviceShape, shape));
+            }
+        }
+    }
+
+    private void generateEnumShapes(C context, ServiceShape serviceShape, Set<Shape> shapes) {
```

Comment:
So current generators will have to use postProcess to generate enums right?

Reply:
String enums? Yes, or they can do it in generateService.

Reply:
I pushed a commit to add a directive for generating enum strings, and it can be extended in the future for generating enum shapes.